File: gadgets/shindig/php/src/socialdata/samplecontainer/XmlStateFileFetcher.php
<?php
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

// State file to use
define('DEFAULT_STATE_FILE', Config::get('base_path') . '/../javascript/samplecontainer/state-basicfriendlist.xml');
// Evil javascript strings
define('REDEFINE_NEW_DATA_REQUEST', "opensocial.newDataRequest = function() { alert('Ha! I attacked you!')}; ");
define('MAKE_PAGE_RED', "document.body.style.backgroundColor = 'red'; ");
define('SCRIPT_PREFIX', "<div onMouseOver=\"" . REDEFINE_NEW_DATA_REQUEST . MAKE_PAGE_RED . "\">");
define('SCRIPT_SUFFIX', "</div>");

class XmlStateFileFetcher {
	private $stateFile;
	private $doEvil = false;
	private $document = null;
	private $allData = null;
	private $friendIdMap = null;
	private $allActivities = null;
	
	// Singleton
	private static $fetcher;
	private function __construct()
	{
		$this->stateFile = DEFAULT_STATE_FILE;
	}
	
	private function __clone()
	{
		// private, don't allow cloning of a singleton
	}
	
	static function get()
	{
		// This object is a singleton
		if (! isset(XmlStateFileFetcher::$fetcher)) {
			XmlStateFileFetcher::$fetcher = new XmlStateFileFetcher();
		}
		return XmlStateFileFetcher::$fetcher;
	}
	
	public function resetStateFile($stateFile)
	{
		$this->stateFile = $stateFile;
		$this->document = null;
		$this->allData = null;
		$this->friendIdMap = null;
		$this->allPeople = null;
		$this->allActivities = null;
	}
	
	public function setEvilness($doEvil)
	{
		$this->doEvil = $doEvil;
	}
	
	private function fetchStateDocument()
	{
		if ($this->document != null) {
			return $this->document;
		}
		// relying on fopen wrappers (which can also handle remote files) and friends
		// since we also want to be able to fwrite new state's in the php version
		// the java version keeps the changes in memory, but since php is multi process
		// instead of single threaded, thats not an option here
		if (($xml = @simplexml_load_file($this->stateFile)) === false) {
			throw new Exception("The state file  {$this->stateFile} could not be fetched and parsed.");
		}
		$this->document = $xml;
		return $this->document;
	}
	
	private function turnEvil($originalString)
	{
		return $this->doEvil ? SCRIPT_PREFIX . $originalString . SCRIPT_SUFFIX : $originalString;
	}
	
	public function getAppData()
	{
		if ($this->allData == null) {
			$this->setupAppData();
		}
		return $this->allData;
	}
	
	private function setupAppData()
	{
		$this->allData = array();
		$xml = $this->fetchStateDocument();
		foreach ( $xml->personAppData->data as $data ) {
			$person = (string)$data['person'];
			$key = (string)$data['field'];
			$value = (string)$data;
			if (!isset($this->allData[$person])) {
				$this->allData[$person] = array();
			}
			$this->allData[$person][$key] = $value;
		}
	}
	
	public function setAppData($id, $key, $value)
	{
		if ($this->allData == null) {
			$this->setupAppData();
		}
		if (! isset($this->allData[$id])) {
			$this->allData[$id] = array();
		}
		$this->allData[$id][$key] = $value;
		
		// Ok to have a fully functioning compliance test in the sample container we need
		// to be able to save data too ... since we don't have shared memory between
		// processes as the java side does, we'll do it the hard way and store the changes
		// in xml.. Oh ps. SimpleXML->addChild() requires PHP >= 5.1.3

		// loop thru the app data to see if this person/key value already exists
		$existingNode = false;
		foreach ( $this->document->personAppData->data as $data ) {
			if ((string)$data['person'] == $id && (string)$data['field'] == $key) {
				$existingNode = $data;
			}
		}
		// if it exists, replace it with the new node, otherwise append the new node
		if ($existingNode !== false) {
			// SimpleXML doesnt have a replace.. so we fall back on dom functions
			$data = new SimpleXMLElement("<data person=\"" . htmlentities($id) . "\" field=\"" . htmlentities($key) . "\">" . htmlentities($value) . "</data>");
			$node1 = dom_import_simplexml($existingNode);
			$dom_sxe = dom_import_simplexml($data);
			$node2 = $node1->ownerDocument->importNode($dom_sxe, true);
			$node1->parentNode->replaceChild($node2, $node1);
		} else {
			// Add a new node with the correct person/field/value to the personData node
			$data = $this->document->personAppData->addChild('data', $value);
			$data->addAttribute('person', $id);
			$data->addAttribute('field', $key);
		}
		if (! @file_put_contents(DEFAULT_STATE_FILE, $this->document->asXML())) {
			throw new Exception("Could not write appData to state file, check the file permissions");
		}
	}
	
	public function getFriendIds()
	{
		if ($this->friendIdMap == null) {
			$this->setupPeopleData();
		}
		return $this->friendIdMap;
	}
	
	public function getAllPeople()
	{
		if ($this->allPeople == null) {
			$this->setupPeopleData();
		}
		return $this->allPeople;
	}
	
	private function setupPeopleData()
	{
		$xml = $this->fetchStateDocument();
		$this->allPeople = array();
		$this->friendIdMap = array();
		
		foreach ( $xml->people->person as $personNode ) {
			$name = (string)$personNode['name'];
			$id = (string)$personNode['id'];
			$person = new Person($id, new Name($this->turnEvil($name)));
			$phoneItem = (string)$personNode['phone'];
			if (! empty($phoneItem)) {
				$phones = array();
				$phones[] = new Phone($this->turnEvil($phoneItem), null);
				$person->setPhoneNumbers($phones);
			}
			$genderItem = (string)$personNode['gender'];
			if (! empty($genderItem)) {
				if ($genderItem == 'F') {
					$person->setGender('FEMALE');
				} else {
					$person->setGender('MALE');
				}
			}
			$this->allPeople[$id] = $person;
			$this->friendIdMap[$id] = $this->getFriends($personNode);
		}
	}
	
	private function getFriends($personNode)
	{
		$friends = array();
		foreach ( $personNode as $friend ) {
			$friend = trim((string)$friend);
			if (! empty($friend)) {
				$friends[] = $friend;
			}
		}
		return $friends;
	}
	
	public function getActivities()
	{
		if ($this->allActivities == null) {
			$this->setupActivities();
		}
		return $this->allActivities;
	}
	
	private function setupActivities()
	{
		$this->allActivities = array();
		$xml = $this->fetchStateDocument();
		foreach ( $xml->activities->stream as $streamItem ) {
			$streamTitle = isset($streamItem['title']) ? (string)$streamItem['title'] : '';
			$userId = isset($streamItem['userId']) ? (string)$streamItem['userId'] : '';
			$this->createActivities($streamItem, $userId, $streamTitle);
		}
	}
	
	private function createActivities($streamItem, $userId, $streamTitle)
	{
		foreach ( $streamItem->activity as $activityItem ) {
			$title = isset($activityItem['title']) ? (string)$activityItem['title'] : '';
			$id = isset($activityItem['id']) ? (string)$activityItem['id'] : 0;
			$body = isset($activityItem['body']) ? (string)$activityItem['body'] : '';
			$activity = new Activity($id, $userId);
			$activity->setStreamTitle($this->turnEvil($streamTitle));
			$activity->setTitle($this->turnEvil($title));
			$activity->setBody($this->turnEvil($body));
			$activity->setMediaItems($this->getMediaItems($activityItem));
			$this->createActivity($userId, $activity);
		}
	}
	
	private function getMediaItems($activityItem)
	{
		$media = array();
		foreach ( $activityItem->mediaItem as $mediaItem ) {
			$type = isset($mediaItem['type']) ? (string)$mediaItem['type'] : '';
			$mimeType = isset($mediaItem['mimeType']) ? $mediaItem['mimeType'] : '';
			$url = isset($mediaItem['url']) ? $mediaItem['url'] : '';
			$media[] = new MediaItem($mimeType, $type, $url);
		}
		return $media;
	}
	
	public function createActivity($userId, $activity)
	{
		
		if ($this->allActivities == null && !is_array($this->allActivities)) {
			$this->setupActivities();
		}
		if (! isset($this->allActivities[$userId])) {
			$this->allActivities[$userId] = array();
		}
		$this->allActivities[$userId][] = $activity;
	}
}

File: gadgets/shindig/java/social-api/src/main/java/org/apache/shindig/social/samplecontainer/XmlStateFileFetcher.java
package org.apache.shindig.social.samplecontainer;

import org.apache.shindig.gadgets.ContentFetcher;
import org.apache.shindig.gadgets.GadgetException;
import org.apache.shindig.gadgets.RemoteContent;
import org.apache.shindig.gadgets.RemoteContentRequest;
import org.apache.shindig.social.opensocial.model.Activity;
import org.apache.shindig.social.opensocial.model.Enum;
import org.apache.shindig.social.opensocial.model.MediaItem;
import org.apache.shindig.social.opensocial.model.Name;
import org.apache.shindig.social.opensocial.model.Person;
import org.apache.shindig.social.opensocial.model.Phone;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;
import java.io.StringReader;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Date;

@Singleton
public class XmlStateFileFetcher {
  private static final String DEFAULT_STATE_URL
      = "http://localhost:8080/gadgets/files/samplecontainer/"
      + "state-basicfriendlist.xml";

  // Evil javascript strings
  private static final String REDEFINE_NEW_DATA_REQUEST
      = "opensocial.newDataRequest = "
      + "function() { alert('Ha! I attacked you!')}; ";

  private static final String MAKE_PAGE_RED
      = "document.body.style.backgroundColor = 'red'; ";

  private static final String SCRIPT_PREFIX = "<div onMouseOver=\""
      + REDEFINE_NEW_DATA_REQUEST + MAKE_PAGE_RED + "\">";

  private static final String SCRIPT_SUFFIX = "</div>";

  private ContentFetcher fetcher;
  private URI stateFile;
  private Document document;
  private boolean doEvil = false;

  // TODO: This obviously won't work on multiple servers
  // If we care then we should do something about it
  private Map<String, Map<String, String>> allData;
  private Map<String, List<String>> friendIdMap;
  private Map<String, Person> allPeople;
  private Map<String, List<Activity>> allActivities;

  @Inject
  public XmlStateFileFetcher(ContentFetcher fetcher) {
   this.fetcher = fetcher;
   try {
      stateFile = new URI(DEFAULT_STATE_URL);
    } catch (URISyntaxException e) {
      throw new RuntimeException(
          "The default state file could not be fetched. ", e);
    }
  }

  public void resetStateFile(URI stateFile) {
    this.stateFile = stateFile;
    document = null;
    allData = null;
    friendIdMap = null;
    allPeople = null;
    allActivities = null;
  }

  public void setEvilness(boolean doEvil) {
    this.doEvil = doEvil;
  }

  private Document fetchStateDocument() {
    if (document != null) {
      return document;
    }

    RemoteContent xml = null;
    try {
      xml = fetcher.fetch(new RemoteContentRequest(stateFile));
    } catch (GadgetException e) {
      throw new RuntimeException(e);
    }

    InputSource is = new InputSource(new StringReader(
        xml.getResponseAsString()));

    DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
    String errorMessage = "The state file " + stateFile
        + " could not be fetched and parsed.";
    try {
      document = factory.newDocumentBuilder().parse(is);
      return document;
    } catch (SAXException e) {
      throw new RuntimeException(errorMessage, e);
    } catch (IOException e) {
      throw new RuntimeException(errorMessage, e);
    } catch (ParserConfigurationException e) {
      throw new RuntimeException(errorMessage, e);
    }
  }

  private String turnEvil(String originalString) {
    if (doEvil) {
     return SCRIPT_PREFIX + originalString + SCRIPT_SUFFIX;
    } else {
      return originalString;
    }
  }

  public Map<String, Map<String, String>> getAppData() {
    if (allData == null) {
      setupAppData();
    }
    return allData;
  }

  private void setupAppData() {
    allData = new HashMap<String, Map<String, String>>();

    Element root = fetchStateDocument().getDocumentElement();

    NodeList elements = root.getElementsByTagName("personAppData");
    NodeList personDataNodes = elements.item(0).getChildNodes();

    for (int i = 0; i < personDataNodes.getLength(); i++) {
      Node personDataNode = personDataNodes.item(i);
      NamedNodeMap attributes = personDataNode.getAttributes();
      if (attributes == null) {
        continue;
      }

      String id = attributes.getNamedItem("person").getNodeValue();
      String field = attributes.getNamedItem("field").getNodeValue();
      String value = personDataNode.getTextContent();

      Map<String, String> currentData = allData.get(id);
      if (currentData == null) {
        currentData = new HashMap<String, String>();
        allData.put(id, currentData);
      }
      currentData.put(field, turnEvil(value));
    }
  }

  public void setAppData(String id, String key, String value) {
    if (allData == null) {
      setupAppData();
    }

    Map<String, String> personData = allData.get(id);
    if (personData == null) {
      personData = new HashMap<String, String>();
      allData.put(id, personData);
    }

    personData.put(key, value);
  }

  public Map<String, List<String>> getFriendIds() {
    if (friendIdMap == null) {
      setupPeopleData();
    }
    return friendIdMap;
  }

  public Map<String, Person> getAllPeople() {
    if (allPeople == null) {
      setupPeopleData();
    }
    return allPeople;
  }

  private void setupPeopleData() {
    Element root = fetchStateDocument().getDocumentElement();

    allPeople = new HashMap<String, Person>();
    friendIdMap = new HashMap<String, List<String>>();
    setupPeopleInXmlTag(root, "people");
  }

  // Adds all people in the xml tag to the allPeople map.
  // Also puts friends ids into the friendIdMap
  private void setupPeopleInXmlTag(Element root, String tagName) {
    // TODO: Use the opensource Collections library
    NodeList elements = root.getElementsByTagName(tagName);
    if (elements == null || elements.item(0) == null) {
      return;
    }

    NodeList personNodes = elements.item(0).getChildNodes();

    for (int i = 0; i < personNodes.getLength(); i++) {
      Node personNode = personNodes.item(i);
      NamedNodeMap attributes = personNode.getAttributes();
      if (attributes == null) {
        continue;
      }

      String name = attributes.getNamedItem("name").getNodeValue();
      String id = attributes.getNamedItem("id").getNodeValue();
      Person person = new Person(id, new Name(turnEvil(name)));
      person.setUpdated(new Date());

      Node phoneItem = attributes.getNamedItem("phone");
      if (phoneItem != null) {
        String phone = phoneItem.getNodeValue();
        List<Phone> phones = new ArrayList<Phone>();
        phones.add(new Phone(turnEvil(phone), null));
        person.setPhoneNumbers(phones);
      }

      Node genderItem = attributes.getNamedItem("gender");
      if (genderItem != null) {
        String gender = genderItem.getNodeValue();
        if ("F".equals(gender)) {
          person.setGender(new Enum<Enum.Gender>(Enum.Gender.FEMALE));
        } else if ("M".equals(gender)) {
          person.setGender(new Enum<Enum.Gender>(Enum.Gender.MALE));
        }
      }

      allPeople.put(id, person);
      friendIdMap.put(id, getFriends(personNode));
    }
  }

  private List<String> getFriends(Node personNode) {
    List<String> friends = new ArrayList<String>();
    NodeList friendNodes = personNode.getChildNodes();
    for (int j = 0; j < friendNodes.getLength(); j++) {
      String friendId = friendNodes.item(j).getTextContent();
      if (friendId != null && friendId.trim().length() != 0) {
        friends.add(friendId.trim());
      }
    }
    return friends;
  }

  public Map<String, List<Activity>> getActivities() {
    if (allActivities == null) {
      setupActivities();
    }

    return allActivities;
  }

  private void setupActivities() {
    allActivities = new HashMap<String, List<Activity>>();

    Element root = fetchStateDocument().getDocumentElement();
    NodeList activitiesElements = root.getElementsByTagName("activities");

    if (activitiesElements != null && activitiesElements.item(0) != null) {
      NodeList streamItems = activitiesElements.item(0).getChildNodes();

      for (int i = 0; i < streamItems.getLength(); i++) {
        Node streamItem = streamItems.item(i);
        NamedNodeMap streamParams = streamItem.getAttributes();
        String streamTitle = "", userId = "";
        if (streamParams != null) {
          streamTitle = streamParams.getNamedItem("title").getNodeValue();
          userId = streamParams.getNamedItem("userId").getNodeValue();
        }

        createActivities(streamItem, userId, streamTitle);
      }
    }
  }

  private void createActivities(Node streamItem, String userId,
      String streamTitle) {
    NodeList activityItems = streamItem.getChildNodes();
    if (activityItems != null) {
      for (int i = 0; i < activityItems.getLength(); i++) {
        Node activityItem = activityItems.item(i);
        NamedNodeMap activityParams = activityItem.getAttributes();
        if (activityParams == null) {
          continue;
        }

        String title = activityParams.getNamedItem("title").getNodeValue();
        String body = activityParams.getNamedItem("body").getNodeValue();
        String id = activityParams.getNamedItem("id").getNodeValue();

        Activity activity = new Activity(id, userId);
        activity.setStreamTitle(turnEvil(streamTitle));
        activity.setTitle(turnEvil(title));
        activity.setBody(turnEvil(body));
        activity.setMediaItems(getMediaItems(activityItem));
        activity.setUpdated(new Date());

        createActivity(userId, activity);
      }
    }
  }

  private List<MediaItem> getMediaItems(Node activityItem) {
    List<MediaItem> media = new ArrayList<MediaItem>();

    NodeList mediaItems = activityItem.getChildNodes();
    if (mediaItems != null) {
      for (int i = 0; i < mediaItems.getLength(); i++) {
        NamedNodeMap mediaParams = mediaItems.item(i).getAttributes();
        if (mediaParams != null) {
          String typeString = mediaParams.getNamedItem("type").getNodeValue();
          String mimeType = mediaParams.getNamedItem("mimeType").getNodeValue();
          String url = mediaParams.getNamedItem("url").getNodeValue();

          media.add(new MediaItem(mimeType,
              MediaItem.Type.valueOf(typeString), url));
        }
      }
    }

    return media;
  }

  public void createActivity(String userId, Activity activity) {
    if (allActivities == null) {
      setupActivities();
    }

    if (allActivities.get(userId) == null) {
      allActivities.put(userId, new ArrayList<Activity>());
    }
    allActivities.get(userId).add(activity);
  }
}
